Fixes #31370 - multi-host stub template snapshots

lzap opened the pull request:
Our snapshot template render test and generator only supports one host stub (ipv4 single NIC redhat OS). In order to test more ERB codepaths, we need a way to create snapshots against multiple host stubs (e.g. IPv6, dualstack, static IPv4 or IPv6, debian family). In order to do that, template metadata can be extended to define list of host stubs to create and test snapshots against.

Snapshot files are now named `ProvisioningTemplate/PXEGrub/Kickstart_default_PXEGrub.host4and6.snap.txt ` where `host4and6` is the stub name, I implemented the following for starters:

* host - the original simple host
* host6 - IPv6 provisioning NIC
* host4and6 - dualstack provisioning NIC
* host4static - single stack IPv4 with static network credentials
* host6static - single stack IPv4 with static network credentials

All stubs are Redhat family OS, we can add more families later (e.g. test debian templates against debian OS). I want this patch to be small, this can be implemented later. I took a chance to remove spaces from snapshot files, they change quite often and `git add` command is now easier.

Finally, there was a suggestion to support dualstack provisioing and while Foreman design is not ideal, after I implemented those snapshot changes I see that our templates do render fine so I am proposing to change the explicit error to just to advertise this via a kernel command line option and leaving the rest up to the users.

https://github.com/theforeman/foreman/pull/8145

lzap commented on app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb:
```diff
@@ -50,7 +50,7 @@ snippet: true
+  elsif subnet6 && subnet6.dhcp_boot_mode? && rhel_compatible && major >= 7
     if host_param_true?("use-slaac")
       options.push("ip=auto6")
     else
```
These changes are coming from https://github.com/theforeman/foreman/pull/8145 and while I'd prefer explicit error, I think it's safer now to allow this since we have a better test coverage.

lzap commented on lib/foreman/renderer/source/snapshot.rb:
```diff
@@ -16,8 +16,8 @@ def snapshot_variants(template)
-            File.join(SNAPSHOTS_DIRECTORY, "#{template_path(template)}.snap.txt")
+          def snapshot_path(template, suffix = 'host')
+            File.join(SNAPSHOTS_DIRECTORY, "#{template_path(template).tr(' ', '_')}.#{suffix}.snap.txt")
           end
```
This change removes spaces from snapshot text files converting them to underscore.

ekohl commented on app/services/foreman/template_snapshot_service.rb:
```diff
@@ -34,33 +72,132 @@ def host
+        ip6: '2001:db8:42::42')
+      domain = FactoryBot.build(:domain, name: 'snapshotdomain.com')
+      subnet6 = FactoryBot.build(:subnet_ipv6, name: 'one', network: '2001:db8:42::')
+      architecture = FactoryBot.build(:architecture, name: 'x86_64')
+      medium = FactoryBot.build(:medium, name: 'CentOS mirror')
+      ptable = FactoryBot.build(:ptable, name: 'ptable')
+      operatingsystem = FactoryBot.build(:operatingsystem, name: 'Redhat',
+                                         major: 7,
+                                         architectures: [architecture],
+                                         media: [medium],
+                                         ptables: [ptable])
+      host = FactoryBot.build(:host, :managed, hostname: 'snapshot-host6',
+                       domain: domain,
+                       subnet6: subnet6,
+                       architecture: architecture,
+                       medium: medium,
+                       ptable: ptable,
+                       operatingsystem: operatingsystem,
+                       interfaces: [interface])
```
This part is mostly the same between all of these. Can we extract that to a helper?

lzap replied:
I prefer verbose version for tests or test stubs. Helpers make code less unreadable in this case.

lzap replied:
Also note I have already extracted some things into `define_host_params` but other lines do not make much sense to me. It's hard to change things than and as I said less readable.

ekohl replied:
Perhaps it could be a [trait in FactoryBot](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#traits)? So you don't define the arch, medium, ptable and OS every time. That already saves a lot of duplication.

lzap replied:
That's a good idea.

ekohl commented on app/views/unattended/provisioning_templates/PXEGrub/kickstart_default_pxegrub.erb:
```diff
@@ -6,6 +6,12 @@ oses:
 - CentOS
 - Fedora
 - RedHat
+test_on:
```
Can we phrase this in a way that's also tells the user something? Perhaps `networking`?

lzap replied:
Well stubs can change anything, not just networking. Better ideas?

ezr-ondrej replied:
Maybe `supported_on`? and make it `ipv4dhcp`, `ipv4and6dhcp`? :thinking:

ekohl commented on app/views/unattended/provisioning_templates/PXEGrub/kickstart_default_pxegrub.erb:
```diff
@@ -6,6 +6,12 @@ oses:
 - Fedora
 - RedHat
+test_on:
+- host
```
Thoughts about renaming this to `host4` or even `host4dhcp`?

lzap replied:
Yeah I was thinking the same, I will do this.

lzap commented:
One more thing this patch could solve is using these "test_on" tags to solve an ugly solution that was introduced some time ago: https://github.com/theforeman/foreman/pull/6951

My thinking is introducing test flag like `with_foreman_ansible` and that would cause the stub to actually return true on `plugin_present?`. This would give us full control on what to actually render during test.

lzap commented:
I was fighting with tests a bit, now I am going to incorporate the comments.

lzap commented:
Tests are green! I actually have one high-priority workaround we need to backport pretty soon which would benefit from this PR merged - I could add a specific test for EL 8.3. Thanks :-)

ronnybremer commented:
Thanks Lukas for all the work you put in.

ezr-ondrej commented on test/factories/host_related.rb:
```diff
@@ -320,6 +320,24 @@ def set_environment_taxonomies(host_or_hostgroup, environment = host_or_hostgrou
       end
     end
 
+    factory :host_for_snapshots_ipv4_dhcp_el7 do
+      name { 'snapshot-ipv4-dhcp-el7' }
+      hostname { name }
+      managed { true }
+      operatingsystem { FactoryBot.build(:for_snapshots_centos_7_0) }
+      domain { FactoryBot.build(:domain_for_snapshots) }
+      subnet { FactoryBot.build(:subnet_ipv4_dhcp_for_snapshots) }
+      pxe_loader { "PXELinux BIOS" }
+      architecture { operatingsystem.try(:architectures).try(:first) }
+      medium { operatingsystem.try(:media).try(:first) }
+      ptable { operatingsystem.try(:ptables).try(:first) }
+      root_pass { '$1$rtd8Ub7R$5Ohzuy8WXlkaK9cA2T1wb0' }
+      environment { FactoryBot.build(:environment, :for_snapshots_test) }
+      certname { name }
+      puppet_proxy { FactoryBot.build(:puppet_smart_proxy, name: 'puppet-proxy', url: 'http://localhost:8448') }
+      puppet_ca_proxy { FactoryBot.build(:puppet_ca_smart_proxy, name: 'puppetca-proxy', url: 'http://localhost:8448') }
+    end
```
can't we just have these as default values in the `TemplateSnapshotService`?
It doesn't really make much sense to have factory, that is only to define some fixed values for one use case and will not be used anywhere else :thinking:

ezr-ondrej reviewed (changes requested):
Thanks for all the hard work @lzap looks great...
Only one complaint in general: I don't like to pollute factories with non-generally useful edge cases, that make it hard to use factories, could we instead set the values in the Snapshot service directly? We can have some method with shared defaults for the host factory, right?

lzap commented:
> Only one complaint in general: I don't like to pollute factories with non-generally useful edge cases

Eh, well me neither. All your comments are valid, however this was asked by @ekohl above. On the other hand, factories directory the right place and I tried to create as descriptive names as possible. So if you ask me, I lean towards keeping them in factories, I can move them back if you want. Whoever is gonna merge my patch I am gonna please :-)

> Maybe supported_on? and make it ipv4dhcp, ipv4and6dhcp

Support is a weird term in this context. I think I like "test_on" better.

Let me know if you insist on moving those fixtures back to the original place. I am rebasing in the meantime.

Note for myself: There are two other PRs that needs to be rebased on top of this.

ezr-ondrej commented:
> On the other hand, factories directory the right place and I tried to create as descriptive names as possible.

I don't agree factories are the right place, but I'm ok with that, if you could just put them in one separate file eg `template_snapshot_related`?

`test_on` is fine by me. It's not perfect, but it works :shrug:

ekohl commented on lib/tasks/snapshots.rake:
```diff
@@ -19,13 +19,14 @@ namespace :snapshots do
-
-          File.open(snapshot_path, 'w') { |f| f.write(snapshot) }
+            snapshot = Foreman::TemplateSnapshotService.render_template(template, host)
+            File.open(snapshot_path, 'w') { |f| f.write(snapshot) }
```
Isn't this the same as
```suggestion
            File.write(snapshot_path, snapshot)
```

ekohl commented on test/factories/domain.rb:
```diff
@@ -14,5 +14,10 @@
     end
+
+    factory :domain_for_snapshots do
+      name { 'snapshotdomain.com' }
+      fullname { 'snapshotdomain.com' }
```
Should we use `example.com`? or `snap.example.com`

lzap replied:
I wanted to differentiate this from exampleN.com but sure. Oh the subdomain is a good idea.

ekohl commented on test/factories/host_related.rb:
```diff
@@ -320,6 +320,24 @@ def set_environment_taxonomies(host_or_hostgroup, environment = host_or_hostgrou
+      name { 'snapshot-ipv4-dhcp-el7' }
+      hostname { name }
+      managed { true }
+      operatingsystem { FactoryBot.build(:for_snapshots_centos_7_0) }
```
This looks like the wrong way to build associations if I read
https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#associations
```suggestion
      operatingsystem factory: :for_snapshots_centos_7_0
```

lzap replied:
Well, that does not work. I know very little about FactoryBot and it took me a while to get this working honestly. Note I am not creating those records, I am building them (not saving them in the db).

I am leaving this as-is feel free to show me a working example.

ekohl commented on test/factories/host_related.rb:
```diff
@@ -320,6 +320,24 @@ def set_environment_taxonomies(host_or_hostgroup, environment = host_or_hostgrou
       end
     end
 
+    factory :host_for_snapshots_ipv4_dhcp_el7 do
```
Why a separate factory? Shouldn't it be a [trait](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#defining-traits)?

lzap replied:
That does not work either, I have no clue why but it looks like for traits some orchestration is being triggered.

Edit: I think `after(:build)` causes troubles. I had to create a new factory for this.

lzap commented:
> I don't agree factories are the right place, but I'm ok with that, if you could just put them in one separate file eg `template_snapshot_related`?

Doesn't this nullify the purpose of a common place? To me this sounds like the worst solution, I'd rather have it in the test itself or in the host factory file.

Honestly, I think there is little to zero chance factories will ever get reused. But it was a huge pain to get it working. If we agree here, I am willing to build those hosts in the test itself like it was originally filed.

lzap commented:
I rebased.

ezr-ondrej commented:
> Honestly, I think there is little to zero chance factories will ever get reused. But it was a huge pain to get it working. If we agree here, I am willing to build those hosts in the test itself like it was originally filed.

I'd lean to having defaults defined in tests as having single purpose factories is quite an antipatern, but it doesn't hurt much eighter, I'm most worried of setting a bad example for the future, but I'm ok to merge anyway if we insist.

ekohl commented:
Since I originally suggested it, let me explain. I suggested to move common patterns to factories. For example, some NIC or OS flavor is used multiple times. That is a good factory/trait. Things that are only used once isn't. So there is some compromise somewhere in between.

ezr-ondrej reviewed (commented):
@lzap can you rebase and just state how you've decide make the data?
I'll merge even if you say you want to keep it in factories :)

ezr-ondrej commented:
@lzap? I believe there is not much left to discuss and whatever you choose to implement I'm fine with that, it increases our test coverage, so I'm all up for that even if it's not implemented in the perfect way.

lzap commented:
Weird, while I am rebasing this patch and regenerating snapshots, I am getting: The snippet 'puppet.conf' threw an error: undefined method '#present?' for Safemode::Jail (Environment) puppet.conf:102:in `bind'

https://github.com/theforeman/foreman/commit/c1104d0046d7d5d53bac7753b9be0b35361d9ef1#diff-3382905c2ea4ecd0ebd292ac3f720b45e89972a2d3f24a6f1d8f1c6586ffc419R21

We do have many `present?` calls, I don't get it. This method should be allowed for years now:

```
commit 0560dacaed5232df6f5ecbc848da608236ab3992
Author: Stephen Benjamin <stbenjam@xxx.com>
Date:   Fri Jan 8 10:51:02 2016 -0500

    Allow any?, blank?, present? in core jails
```

lzap commented:
What is weird is it only fails for the `puppet.conf` snippet, we have many other occurrences and these work fine.

lzap commented:
I think I have found the problem, method `present?` is allowed in Ruby core Object in safemode, however it is also redefined in ActiveRecord::Core. And in rendering tests whole Environment instance is stubbed instead of String.

I think it is safe to allow `present?` method from ActiveRecord, it returns always true anyway. I am doing this in the ActiveRecord base class in Foreman because I think this is where it belongs - we do have an apipie documentation group that documents allowing of id and name however this is actually not done so the documentation is misleading.

lzap commented:
There is one more issue I need to solve, unattended_url setting is set in fixtures to `foreman.some.host.fqdn` but in `settings.yaml.test` to `foreman.example.com` and this creates problem in snapshot tests. While I could override the value, I think the cleaner approach would be to simple fix everything to the preferred "example.com" domain which is often recommended by @ekohl - so let me do that in this PR, tests will not pass until this is done.

Note for me - the commit before this change is eae4dc3626d10fffa081138739ca646d7800668a

ezr-ondrej commented on test/fixtures/settings.yml:
```diff
@@ -3,12 +3,12 @@ attributes1:
   name: foreman_url
   category: Setting::General
-  default: http://foreman.some.host.fqdn
+  default: http://foreman.example.com
```
:heart:

ezr-ondrej commented:
Though I agree changing the domain is a good way to go, I'd prefer it to be at least in separate commit, maybe even PR for better visibility and cleaner history. Also this should align with the Setting changes, where I'll need to touch these as well as fixtures won't be viable to change the default value anymore.

lzap commented:
> Though I agree changing the domain is a good way to go, I'd prefer it to be at least in separate commit, maybe even PR for better visibility and cleaner history. Also this should align with the Setting changes, where I'll need to touch these as well as fixtures won't be viable to change the default value anymore.

Do you insist on this? Thing is that the new factories already use domain `example.com` and I thought it is a good time to actually change everything to example.com. I'd probably need to change everything to that weird some.host.fqdn (including the new factories) only after I would change it back in another PR.

lzap commented:
I fixed the template change which was caused by having Ansible plugin. There is a PR to fix that scheduled to be merged after this one is merged :-)

ezr-ondrej commented:
[test katello]

ezr-ondrej reviewed (dismissed):
I believe this is worth merging even though we would cause trouble to stabilization, and it is not much chance to cause trouble as it touches mostly just tests.
But OTH, nothing is really forcing us to get it in 2.5 IIUC.
@tbrisker ? :)

lzap commented:
I would love to get this in, I spent quite time on this and it vastly improves test coverage of templates.

I am rebasing due to conflict caused by the other PR.

tbrisker commented:
@ezr-ondrej @lzap i'm fine with merging this during stabilization unless there is risk this could cause plugin tests to fail - in which case, let's wait a week and give plugin maintainers time to make sure their plugins are working against 2.5

lzap commented:
Marking my calendar then, week it is, thanks :-)

lzap commented:
I think I got it incorrectly, I do not thing this should introduce any problems to plugin authors. It just adds new contexts to the template renderer tests. There should not be any risk I am aware of.

tbrisker commented on app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb:
```diff
@@ -32,7 +32,7 @@ snippet: true
   options.push("kssendmac", "ks.sendmac", "inst.ks.sendmac")
 
   # networking credentials
-  raise("Dual-stack provisioning not supported") if subnet4 && subnet6
```
This and the other change to this file do not seem like they are related to this issue and probably deserve their own separate PR

lzap replied:
Ah this is a leftover from https://github.com/theforeman/foreman/pull/8447 that should not be here for sure.

ezr-ondrej replied:
can you remove it then? :)

tbrisker commented on app/models/application_record.rb:
```diff
@@ -8,6 +8,11 @@ class ApplicationRecord < ActiveRecord::Base
+  end
+
+  class Jail < Safemode::Jail
+    allow :id, :name, :present?
```
Also not related change, looks like `:name` comes from safemode itself: https://github.com/svenfuchs/safemode/blob/master/lib/safemode/core_jails.rb#L127
and i'm not sure about `:id` but it seems that it is already allowed from somewhere?

lzap replied:
I could not find where exactly `id` is allowed, the point of this hunk is that this class has documentation that says these are all allowed (I also added present? on top of that which is safe to add). Problem is, it was unclear where it gets added. So I wanted to explicitly name all three here, Safemode seems to walk up the class hierarchy when finding allowed methods and it is okay to "override" them.

Do you prefer not to mention `name` since it is allowed for all Ruby objects? For the record, we do actually allow it in the individual model classes anyway, so this is not a deviation from our approach. I prefer clean Jail here which matches the documentation rather than partial list to be honest.

lzap committed:
Fixes #31370 - multi-host stub template snapshots

lzap commented:
Removed that deleted dualstack line. Also removed the `foreman.some.host.fqdn` to `foreman.example.com` change as this breaks plugins and it is not worth doing this in this PR.

lzap commented:
I fixed the tests

lzap commented:
And it is FINALLY green!

ezr-ondrej reviewed (approved):
Thanks @lzap ! :+1:
